### ISSUE TITLE
Odata Bind uses EntityType not EntitySet

### ISSUE THREAD

**Issue by Fridgenz** [object/Simple.OData.Client#511]
Possibly just a question, possibly a bug.

Given some simplified metadata:

<EntityType Name="XResource">
	<Key>
		<PropertyRef Name="Id"/>
	</Key>
	<Property Name="Id" Type="Edm.Guid" Nullable="false"/>
	<NavigationProperty Name="Y" Type="Models.YResource"/>
</EntityType>

<EntitySet Name="Xs" EntityType="Models.XResource">

<EntityType Name="YResource">
	<Key>
		<PropertyRef Name="Id"/>
	</Key>
	<Property Name="Id" Type="Edm.Guid" Nullable="false"/>
</EntityType>

<EntitySet Name="Ys" EntityType="Models.YResource">

Odata Service endpoints:

service/v2/Xs(Guid) returns a XResource
service/v2/Ys(Guid) returns a YResource

Attempting to create an XResource entry:

Client.For<OdataServiceX>("Xs")
	.Set(backendX)
	.InsertEntryAsync()

where C#:

backendX = BackendX { y = BackendY(Guid) } 

generates odata body:

XResource {
	Y@odata.bind=service/v2/YResource(Guid)
}

where I would expect

XResource {
	Y@odata.bind=service/v2/Ys(Guid)
}

Changing Type name of "BackendY" to "Y" will make the client correctly find service/v2/Ys(Guid)

It seems the client uses the EntityType and not the EntitySet for trying to best match an unknown type name?
Is there a way to use classes that are not named the same as the model endpoint names?

**Comment by object** [comment]
I agree that EntitySet name should be preferred to EntityType, I will check this out. As a workaround you can attach either Table or DataContract attribute to a class, e.g.

[Table(Name="Ys")]
public class OdataServiceY

**Comment by Fridgenz** [comment]
The annotations didn't seem to work for me. I tried both Table and DataContract on the class and Column on the property.

What did work, is pulling the source into my project and adding: 

```
if (TryGetEntityType(typeName, out entityType) && TryGetEntitySetByType(entityType.FullTypeName(), out entitySet))
{
    return entitySet.Name;
}
```
To metadata.cs just before the entityType.Name is used (line 124).

```
private bool TryGetEntitySetByType(string entityType, out IEdmEntitySet entitySet) {
    if (entityType.Contains("/"))
	    entityType = entityType.Split('/').First();

    entitySet = _model.SchemaElements
	    .Where(x => x.SchemaElementKind == EdmSchemaElementKind.EntityContainer)
	    .SelectMany(x => (x as IEdmEntityContainer).EntitySets())
	    .BestMatch(x => x.Type.AsElementType().FullTypeName(), entityType, _session.Settings.NameMatchResolver);

    return entitySet != null;
}
```

Does this solution break logic elsewhere?

**Comment by object** [comment]
Thank you for your finding. I will investigate this further.

**Comment by JaxomF** [comment]
Hi,
Any update on this bug ?

**Comment by CJTalbot87** [comment]
I have also have this issue where the EntityType is used and not the EntitySet, the DataAnnotations also did not work for me.

Is there any "clean" way to override the used OData Bind entity name?

**Comment by swidz** [comment]
I am having also this issue, I have not digged to the source code yet, but is behaves strange. For most of the odata.bindings the entitySet is used but for one which actually causes the problem a type name is used.
